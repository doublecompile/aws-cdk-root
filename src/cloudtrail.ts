import { Stack, Duration } from "aws-cdk-lib";
import { CfnTrail, Trail } from "aws-cdk-lib/aws-cloudtrail";
import { ServicePrincipal, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { Key } from "aws-cdk-lib/aws-kms";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import {
  Bucket,
  BlockPublicAccess,
  BucketEncryption,
  StorageClass,
} from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

/**
 * Constructor properties for CloudTrailSetup.
 */
export interface CloudTrailSetupProps {
  /**
   * The ID of the orginaziton in AWS Organizations
   */
  readonly organizationId?: string;

  /**
   * The number of days the log group may retain logs.
   */
  readonly logGroupRetentionDays?: number;
}

/**
 * The setup for CloudTrail.
 */
export class CloudTrailSetup extends Construct {
  /**
   * The S3 bucket.
   */
  public readonly bucket: Bucket;

  /**
   * The CloudTrail trail construct.
   */
  public readonly trail: Trail;

  /**
   * The CloudWatch Logs group.
   */
  public readonly logGroup: LogGroup;

  /**
   * The KMS key for encryption at rest.
   */
  public readonly encryptionKey: Key;

  /**
   * @param scope - The scope in which to define this construct.
   * @param id - The scoped construct ID.
   * @param props - The properties.
   * @param props.organizationId - The AWS Organizations ID of the root account.
   * @param props.logGroupRetentionDays - The number of days to keep CloudWatch Logs entries.
   */
  constructor(scope: Construct, id: string, props: CloudTrailSetupProps = {}) {
    super(scope, id);

    const { region, account } = Stack.of(this);

    const { organizationId, logGroupRetentionDays } = props;

    if (!organizationId) {
      throw new Error('"organizationId" is required');
    }

    const cloudTrailPrincipal = new ServicePrincipal(
      "cloudtrail.amazonaws.com"
    );

    const storageKey = new Key(this, "Key", {
      description: "Encrypts CloudTrail destination S3 bucket",
      enabled: true,
      enableKeyRotation: true,
    });
    this.encryptionKey = storageKey;

    this.bucket = new Bucket(this, "Storage", {
      encryption: BucketEncryption.KMS,
      encryptionKey: storageKey,
      bucketKeyEnabled: true,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    });
    this.bucket.addLifecycleRule({
      transitions: [
        {
          storageClass: StorageClass.INFREQUENT_ACCESS,
          transitionAfter: Duration.days(30),
        },
      ],
    });

    // While the CDK takes care of PutObject for non-organization trails, we
    // must add our own policy for the organization trail.
    const bucketOrganizationPolicy = new PolicyStatement({
      resources: [
        this.bucket.arnForObjects(`AWSLogs/${organizationId}/${account}/*`),
      ],
      actions: ["s3:PutObject"],
      principals: [cloudTrailPrincipal],
      conditions: {
        StringEquals: { "s3:x-amz-acl": "bucket-owner-full-control" },
      },
    });
    this.bucket.addToResourcePolicy(bucketOrganizationPolicy);

    // Per the AWS documentation:
    // https://docs.aws.amazon.com/awscloudtrail/latest/userguide/create-kms-key-policy-for-cloudtrail.html
    // https://docs.aws.amazon.com/awscloudtrail/latest/userguide/encrypting-cloudtrail-log-files-with-aws-kms.html
    // CloudTrail must be granted DescribeKey, Decrypt, and GenerateDataKey*.
    const cloudTrailStorageKeyGrant =
      storageKey.grantEncryptDecrypt(cloudTrailPrincipal);
    for (const statement of cloudTrailStorageKeyGrant.resourceStatements) {
      statement.addActions("kms:DescribeKey");
    }

    this.logGroup = new LogGroup(this, "LogGroup", {
      encryptionKey: storageKey,
      retention: logGroupRetentionDays || RetentionDays.ONE_MONTH,
    });

    // Per the AWS documentation:
    // https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/encrypt-log-data-kms.html
    // CloudWatch Logs must be granted Encrypt/Decrypt and Describe*.
    const logGroupPrincipal = new ServicePrincipal(
      `logs.${region}.amazonaws.com`
    );
    const cloudWatchLogsKeyGrant =
      storageKey.grantEncryptDecrypt(logGroupPrincipal);
    for (const statement of cloudWatchLogsKeyGrant.resourceStatements) {
      statement.addActions("kms:Describe*");
    }

    this.trail = new Trail(this, "Trail", {
      encryptionKey: storageKey,
      bucket: this.bucket,
      sendToCloudWatchLogs: true,
      cloudWatchLogGroup: this.logGroup,
      isMultiRegionTrail: true,
      // You have to name the trail in order to use this property on the L2
      // isOrganizationTrail: true,
      // orgId: organizationId,
      // trailName: "NameOfYourTrail",
    });
    const cfnTrail = this.trail.node.defaultChild as CfnTrail;
    cfnTrail.addPropertyOverride("IsOrganizationTrail", true);

    for (const statement of cloudTrailStorageKeyGrant.resourceStatements) {
      statement.addCondition("StringLike", {
        "aws:SourceArn": `arn:aws:cloudtrail:${region}:${account}:trail/*`,
      });
    }
    for (const statement of cloudWatchLogsKeyGrant.resourceStatements) {
      statement.addCondition("ArnLike", {
        "kms:EncryptionContext:aws:logs:arn": `arn:aws:logs:${region}:${account}:log-group:*`,
      });
    }

    // For limiting the key usage to specific accounts.
    //   "StringLike": {
    //     "kms:EncryptionContext:aws:cloudtrail:arn": [
    //       "arn:aws:cloudtrail:*:account-id:trail/*"
    //     ]
    //   },
  }
}
