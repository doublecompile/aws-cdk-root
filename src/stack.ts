import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { CloudTrailSetup } from "./cloudtrail";

export class RootAccountStack extends Stack {
  /**
   *
   * @param scope - The parent construct
   * @param id - The construct ID
   * @param props - The construct properties
   */
  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props);

    new CloudTrailSetup(this, "AuditLog", {
      organizationId: this.node.tryGetContext("OrganizationId"),
    });
  }
}
