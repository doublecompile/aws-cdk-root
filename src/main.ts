import { App } from "aws-cdk-lib";
import { RootAccountStack } from "./stack";

// for development, use account/region from CDK CLI
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

// Construct our AWS CDK app.
const app = new App();

// Instantiate the RootAccountStack.
new RootAccountStack(app, "OrgRoot", {
  env: devEnv,
});

// Synthesize the CloudFormation template.
app.synth();
