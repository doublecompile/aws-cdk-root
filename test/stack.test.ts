import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { RootAccountStack } from "../src/stack";

test("Snapshot", () => {
  const app = new App({ context: { OrganizationId: "o-123" } });
  const stack = new RootAccountStack(app, "test", {});

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
