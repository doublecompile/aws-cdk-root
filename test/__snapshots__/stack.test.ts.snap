// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AuditLogLogGroupCEC160CB": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "AuditLogLogGroupKey169DA879",
            "Arn",
          ],
        },
        "RetentionInDays": 30,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "AuditLogLogGroupKey169DA879": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Description": "Encrypts CloudTrail destination CloudWatch Log Group",
        "EnableKeyRotation": true,
        "Enabled": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
                "kms:Describe*",
              ],
              "Condition": {
                "ArnLike": {
                  "kms:EncryptionContext:aws:logs:arn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:logs:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:*",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::Join": [
                    "",
                    [
                      "logs.",
                      {
                        "Ref": "AWS::Region",
                      },
                      ".amazonaws.com",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "AuditLogStorage9B1900A2": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "BucketKeyEnabled": true,
              "ServerSideEncryptionByDefault": {
                "KMSMasterKeyID": {
                  "Fn::GetAtt": [
                    "AuditLogStorageKeyD95120E5",
                    "Arn",
                  ],
                },
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "LifecycleConfiguration": {
          "Rules": [
            {
              "Status": "Enabled",
              "Transitions": [
                {
                  "StorageClass": "STANDARD_IA",
                  "TransitionInDays": 30,
                },
              ],
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "AuditLogStorageKeyD95120E5": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "Description": "Encrypts CloudTrail destination S3 bucket",
        "EnableKeyRotation": true,
        "Enabled": true,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            {
              "Action": [
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
                "kms:DescribeKey",
              ],
              "Condition": {
                "StringLike": {
                  "aws:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:aws:cloudtrail:",
                        {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":trail/*",
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "AuditLogStoragePolicy5919A257": {
      "Properties": {
        "Bucket": {
          "Ref": "AuditLogStorage9B1900A2",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutObject",
              "Condition": {
                "StringEquals": {
                  "s3:x-amz-acl": "bucket-owner-full-control",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "AuditLogStorage9B1900A2",
                        "Arn",
                      ],
                    },
                    "/AWSLogs/o-123/",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    "/*",
                  ],
                ],
              },
            },
            {
              "Action": "s3:GetBucketAcl",
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "AuditLogStorage9B1900A2",
                  "Arn",
                ],
              },
            },
            {
              "Action": "s3:PutObject",
              "Condition": {
                "StringEquals": {
                  "s3:x-amz-acl": "bucket-owner-full-control",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudtrail.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "AuditLogStorage9B1900A2",
                        "Arn",
                      ],
                    },
                    "/AWSLogs/",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "AuditLogTrailEB39A870": {
      "DependsOn": [
        "AuditLogStoragePolicy5919A257",
        "AuditLogTrailLogsRoleDefaultPolicy1FE1E3B2",
        "AuditLogTrailLogsRole5C997FC2",
      ],
      "Properties": {
        "CloudWatchLogsLogGroupArn": {
          "Fn::GetAtt": [
            "AuditLogLogGroupCEC160CB",
            "Arn",
          ],
        },
        "CloudWatchLogsRoleArn": {
          "Fn::GetAtt": [
            "AuditLogTrailLogsRole5C997FC2",
            "Arn",
          ],
        },
        "EnableLogFileValidation": true,
        "EventSelectors": [],
        "IncludeGlobalServiceEvents": true,
        "IsLogging": true,
        "IsMultiRegionTrail": true,
        "IsOrganizationTrail": true,
        "KMSKeyId": {
          "Fn::GetAtt": [
            "AuditLogStorageKeyD95120E5",
            "Arn",
          ],
        },
        "S3BucketName": {
          "Ref": "AuditLogStorage9B1900A2",
        },
      },
      "Type": "AWS::CloudTrail::Trail",
    },
    "AuditLogTrailLogsRole5C997FC2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudtrail.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "AuditLogTrailLogsRoleDefaultPolicy1FE1E3B2": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutLogEvents",
                "logs:CreateLogStream",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "AuditLogLogGroupCEC160CB",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AuditLogTrailLogsRoleDefaultPolicy1FE1E3B2",
        "Roles": [
          {
            "Ref": "AuditLogTrailLogsRole5C997FC2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
